actionpack/lib/action_dispatch/http/filter_redirect.rb

```diff
@@ -22,13 +20,11 @@ def location_filters
         end
       end
 
-      def location_filter_match?
-        location_filters.any? do |filter|
-          if String === filter
-            location.include?(filter)
-          elsif Regexp === filter
-            location =~ filter
-          end
+      KV_RE   = "[^&;=]+"
+      PAIR_RE = %r{(#{KV_RE})=(#{KV_RE})}
```

Comment:
These constants aren't needed. I think it might make more sense to parse the location into a URI and parse the query into a hash using Rack.

eg.
```
def filtered_location # :nodoc:
  uri = URI.parse(location)
  uri.to_s.tap do |filtered|
    filtered.sub!(uri.host, location_filter(uri.host))
    filtered.sub!(uri.path, location_filter(uri.path))
    filtered.sub!(uri.query, location_parameter_filter(uri.query)) if uri.query
  end
end

def location_parameter_filter(params)
  filter = ActiveSupport::ParameterFilter.new(location_filters)
  filter.filter(Rack::Utils.parse_query(params)).to_query.presence
end
```